README.md
# [Power BI](https://powerbi.microsoft.com/) [R Script](https://cran.r-project.org/) Data Sources for Hornbill Reporting And Trend Engine

## Overview

These example scripts have been provided to enable Power BI administrators to build reports and dashboards using Hornbill Reporting and Advanced Analytics Trend Engine data as their data source(s).

## Dependencies

The scripts have been written in [R](https://cran.r-project.org/), and were developed using the following:

- [Power BI Desktop build 2.75.5649.861 64-bit (November 2019)](https://powerbi.microsoft.com/)
- [Microsoft R Open 3.5.3](https://mran.microsoft.com/open/)

The following packages are required dependencies, and can be installed via the CRAN repositories:

- [httr](https://cran.r-project.org/web/packages/httr/)
- [data.table](https://cran.r-project.org/web/packages/data.table/) - Just for the TrendingData script
- [readxl](https://cran.r-project.org/web/packages/readxl/) - Just for the Report and HistoricReport scripts, when useXLSX is set to TRUE

## Configuration used in all scripts

Each script requires the following variables to be set (all case-sensitive):

- instanceName - This is the name of the instance to connect to. So for instance, if you use https://live.hornbill.com/yourinstancename to connect to Hornbill, then **yourinstancename** is the part of the URL that should be used in this variable. Note, this is case sensitive.
- apiKey - This is an API key generated against a user account on the Hornbill Administration Console, where the user account has sufficient access to run reports and access trending data.

Each script an be configured to use a proxy for access to your Hornbill instance. Set all of the below to NULL to not use a proxy. If using a proxy, the proxyAddress and proxyPort are the minimum required to be provided.

- proxyAddress - The hostname or IP address of the proxy
- proxyPort - The proxy port
- proxyUsername - The username to access the proxy, if required
- proxyPassword - The password for the above account
- proxyAuth - The type of HTTP authentication to use. Should be one of the following: basic, digest, digest_ie, gssnegotiate, ntlm, any.

## Scripts

### PowerBIDataSource_Report.R

This script will:

- Run a pre-defined report on the Hornbill instance;
- Wait for the report to complete;
- Retrieve the report CSV data and present back as an R data frame called output, which can then be retrieved and reported on by PowerBI.

Script Variables:

- reportID: The ID (Primary Key, INT) of the  report to be run;
- reportComment: A comment to write against the report run job.
- deleteReportInstance: a boolean value to determine if, once the report is run on Hornbill and the data has been pulled in to PowerBI, whether the historic report run instance should be removed from your Hornbill report.
- useXLSX: FALSE = the script will use the CSV output from your report; TRUE = the script will use the XLSX output from your report. NOTE: XLSX output will need to be enabled within the Output Formats > Additional Data Formats section of your report in Hornbill;
- deleteLocalXLSX: FALSE = the downloaded XLSX file will remain on disk once the extract is complete; TRUE = the local XLSX file is deleted upon completion
- xlsxLocalFolder: The folder where to store the downloaded XLSX file. Can be left blank, or specify a local folder to store the downloaded XLSX file into. Requires the postfixed / or \ on the path, depending on your OS
- csvEncoding: The character set to be used when decoding the CSV report data, or when converting the XLSX data into a Power BI friendly codepage. This will usually be "UTF-8", but if you have issues returning data with certain characters (the Windows E2 80* characters are the usual culprits) then choose a different character set to use, ie: "ISO-8859-1". Look out for an error that looks like this for character set issues: "Details: "Unable to translate bytes [E2][80] at index 1077 from specified code page to Unicode"".
- suspendSeconds: The number of seconds the script should wait between checks to see if the report is complete. NOTE : there is a defect/incompatibility between Power BI and the RCurl library that we are using to make the HTTP requests to Hornbill, where if more than 4 or 5 calls with getURL are made within the same script then getURL hangs until Power BI releases it. Increasing the number of seconds between checks reduces the required number of calls to your Hornbill instance, and will fix data source hanging issues.

### PowerBIDataSource_HistoricReport.R

This script will:

- Retrieve a historic report CSV from your Hornbill instance;
- Present the report data back as an R data frame called dataframe, which can then be retrieved and reported on by PowerBI.

Script Variables:

- reportID: The ID (Primary Key, INT) of the  report to be run;
- runId: The Run ID (INT) of a historic run of the above report ID.
- useXLSX: FALSE = the script will use the CSV output from your report; TRUE = the script will use the XLSX output from your report. NOTE: XLSX output will need to be enabled within the Output Formats > Additional Data Formats section of your report in Hornbill;
- deleteLocalXLSX: FALSE = the downloaded XLSX file will remain on disk once the extract is complete; TRUE = the local XLSX file is deleted upon completion
- xlsxLocalFolder: The folder where to store the downloaded XLSX file. Can be left blank, or specify a local folder to store the downloaded XLSX file into. Requires the postfixed / or \ on the path, depending on your OS
- csvEncoding: The character set to be used when decoding the CSV report data, or when converting the XLSX data into a Power BI friendly codepage. This will usually be "UTF-8", but if you have issues returning data with certain characters (the Windows E2 80* characters are the usual culprits) then choose a different character set to use, ie: "ISO-8859-1". Look out for an error that looks like this for character set issues: "Details: "Unable to translate bytes [E2][80] at index 1077 from specified code page to Unicode"".

### PowerBIDataSource_TrendingData.R

This script will:

- Run the reporting::measureGetInfo API  against your Hornbill instance, with a given measure ID (Primary Key, INT);
- Build a table containing all Trend Value entries for the selected measure;
- Present the trend data back as an R data frame called dataframe, which can then be retrieved and reported on by PowerBI.

Script Variable:

- measureID: The ID (Primary Key, INT) of the measure to return trend data from.

Outputs:
As the response parameters from the Trending Engine is fixed (unlike the Reporting engine, which has user-specified column outputs), the output for this report will always consist  of the following columns:

- value - the value of the trend sample;
- sampleId - the ID of the sample;
- sampleTime - the time & date that the sample was taken;
- dateRange.from - the start date of the sample snapshot;
- dateRange.to - the end date of the sample snapshot;

## Power BI Notes

These scripts have been designed to be used as data sources only, and not as the source of R script visuals within Power BI. Which is not to say they couldn't be used in your R script visuals, with a little extra code :)


CHANGELOG.md
# CHANGELOG

## 1.8.3 (Dec 8th, 2023)

Changes:

- Updated XMLMC API Endpoint
- Inclided Dav Endpoint

## 1.8.2 (Just 2nd, 2021)

Changes:

- Updated method to pull CSV report data

## 1.8.1 (June 30th, 2021)

Changes:

- URL encoded the report filename 

## 1.8.0 (November 12th, 2020)

Features:

- Added support for data in XLSX files to be added to the output dataframe in the codepage as defined in the csvEncoding variable 

## 1.7.0 (November 12th, 2020)

Features:

- Added support for XLSX report output for Report and HistoricReport scripts

## 1.6.0 (November 9th, 2020)

Features:

- Added support for updated JSON API response
- Removed jsonlite and readr dependency
- Introduced new dependency in TrendingData script - [data.table](https://cran.r-project.org/web/packages/data.table/)

## 1.5.0 (June 29th, 2020)

Features:

- Added support for proxies

## 1.4.0 (January 20th, 2020)

Features:

- Better handling and reporting of API call errors

## 1.3.0 (December 30th, 2019)

Features:

- Replaced textConnection & CSV reader code, replaced with httr content parser
- Added support to define which character set to use when parsing CSV code

## 1.2.0 (November 26th, 2019)

Features:

- Added support for TLS 1.2/1.3
- Removed need to provide instance zone
- General tidy-up of code
- Updated package dependencies - now requires httr and jsonlite, and no longer requires RCurl and XML

## 1.1.0 (June 28th, 2018)

Features:
- Added functionality to delete the historic report run instance from Hornbill once the data has been imported to Power BI;
- Added suspend code to work around a Power BI / RCurl incompatibility, where if you make more than 4 HTTP calls using getURL in the same script, the script hangs until you hit cancel in Power BI.

## 1.0.0 (May 9th, 2017)

Features:
- Initial Release


PowerBIDataSource_Report.R
#Define Instance Details
instanceName = "yourinstanceid"
apiKey = "yourapikey"

# Define Report details
reportID = "1"
reportComment = "A comment to add to the report run"
deleteReportInstance <- FALSE
useXLSX <- TRUE # FALSE = the script will use the CSV output from your report; TRUE = the script will use the XLSX output from your report

# Settings for using XLSX report output
# You must have XLSX output enabled against the target report in your Hornbill instance to use these settings
deleteLocalXLSX <- FALSE  # FALSE = the downloaded XLSX file will remain on disk once the extract is complete; TRUE = the local XLSX file is deleted upon completion
xlsxLocalFolder <- "" # Can be left blank, or specify a local folder to store the downloaded XLSX file into. Requires the postfixed / or \ depending on your OS

# Settings for using CSV report output, or character encoding conversion for XLSX repost output
csvEncoding <- "UTF-8"    # For Unicode byte translation issues in Power BI, try using ISO-8859-1 as the value for csvEncoding

# Suspend for X amount of seconds between checks to see if the report is complete
suspendSeconds <- 1

# Define Proxy Details
proxyAddress <- NULL # "127.0.0.1" - location of proxy
proxyPort <- NULL # 8080 - proxy port
proxyUsername = NULL # login details for proxy, if needed
proxyPassword = NULL # login details for proxy, if needed
proxyAuth = NULL # "any" - type of HTTP authentication to use. Should be one of the following: basic, digest, digest_ie, gssnegotiate, ntlm, any.

# Import dependencies
library(httr)

# Set httr default timeout, defaults to 10 seconds
set_config( config( connecttimeout = 60 ) )

# Get Endpoint
responseFromFiles <- GET(paste("https://files.hornbill.com/instances", instanceName, "zoneinfo",sep="/"), 
                         use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                         add_headers('Content-Type'='application/json',Accept='application/json'))

xmlmcURL <-  content(responseFromFiles, as = "parsed", type = "application/json", encoding="UTF-8")$zoneinfo$apiEndpoint
davURL <-  content(responseFromFiles, as = "parsed", type = "application/json", encoding="UTF-8")$zoneinfo$davEndpoint

# invokeXmlmc - take params, fire off XMLMC call
invokeXmlmc = function(service, xmethod, params) {
  paramsrequest = paste(params , collapse="")
  arrRequest = c(	"<methodCall service=\"",
                  service, 
                  "\" method=\"",
                  xmethod,
                  "\">",
                  "<params>",
                  paramsrequest,
                  "</params>",
                  "</methodCall>")
  responseFromURL <- POST(paste(xmlmcURL, service, "?method=", xmethod, sep=""),
                          add_headers('Content-Type'='text/xmlmc',Accept='application/json', Authorization=paste('ESP-APIKEY', apiKey, sep=" ")),
                          use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                          body=paste(arrRequest, collapse=""))
  return(responseFromURL)
}

# Kick off report run, get job ID
runOutput <- content(invokeXmlmc("reporting", "reportRun", paste("<reportId>", reportID, "</reportId>","<comment>", reportComment, "</comment>")), encoding="UTF-8")
runStatus <- runOutput$"@status"

if (runStatus == FALSE || runStatus == "fail") {
  stop(runOutput$state$error)
} else {
  
  runID <- runOutput$params$runId
  
  reportSuccess <- FALSE
  reportComplete <- FALSE
  
  if (runID > 0) {
    repeat {
      Sys.sleep(suspendSeconds)
      
      # Check status of report
      reportRunStatus <- invokeXmlmc("reporting", "reportRunGetStatus", paste("<runId>", runID, "</runId>"))
      runStatus <- content(reportRunStatus)$params$reportRun$status
      runComp <- grepl(runStatus, "completed")
      
      if (runComp == TRUE) {
        for (file in content(reportRunStatus)$params$files) {
          if (file$type == "xlsx" && useXLSX == TRUE) {
            reportLink = file$name
          } else if (file$type == "csv" && useXLSX == FALSE) {
            reportLink = file$name
          }
        }
        reportSuccess <- TRUE
        reportComplete <- TRUE
        break;
      } else if (runStatus == "failed") {	
        reportSuccess <- FALSE
        reportComplete <- TRUE
        break;
      }
    }
  }
  
  if (reportSuccess == FALSE) {	
    stop()
  }
  
  # GET request for report CSV content
  reportLinkLocal <- paste(xlsxLocalFolder, reportLink, sep="")
  reportContent <- GET(paste(davURL,"reports", reportID, URLencode(reportLink), sep="/"),
                       write_disk(reportLinkLocal, overwrite=TRUE),
                       use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                       add_headers('Content-Type'='text/dav', Authorization=paste('ESP-APIKEY ', apiKey, sep="")))
  # Delete the report run instance  
  if (deleteReportInstance == TRUE) {
    reportDeleteHist <- invokeXmlmc("reporting", "reportRunDelete", paste("<runId>", runID, "</runId>"))
  }
  
  # vector into data frame object
  if (useXLSX == TRUE) {
    library(readxl)
    write.csv(read_excel(reportLinkLocal),paste(reportLinkLocal, "csv", sep="."), row.names = FALSE)
    output <-read.csv(paste(reportLinkLocal, "csv", sep="."), encoding = csvEncoding, stringsAsFactors=FALSE)
    if (deleteLocalXLSX == TRUE && file.exists(reportLinkLocal)) {
      file.remove(reportLinkLocal)
    }
    if (file.exists(paste(reportLinkLocal, "csv", sep="."))) {
      file.remove(paste(reportLinkLocal, "csv", sep="."))
    }
  } else {
    output <- content(reportContent, as = "parsed", type = "text/csv", encoding = csvEncoding)
  }
}


PowerBIDataSource_TrendingData.R
#Define Instance Details
instanceName = "yourinstanceid"
apiKey = "yourapikey"

# Define Measure details
measureID = "10"

# Define Proxy Details
proxyAddress <- NULL # "127.0.0.1" - location of proxy
proxyPort <- NULL # 8080 - proxy port
proxyUsername = NULL # login details for proxy, if needed
proxyPassword = NULL # login details for proxy, if needed
proxyAuth = NULL # "any" - type of HTTP authentication to use. Should be one of the following: basic, digest, digest_ie, gssnegotiate, ntlm, any.

# Import dependencies
library('httr')
library(data.table)

# Set httr default timeout, defaults to 10 seconds
set_config( config( connecttimeout = 60 ) )

# Get Endpoint
responseFromFiles <- GET(paste("https://files.hornbill.com/instances", instanceName, "zoneinfo",sep="/"), 
                         use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                         add_headers('Content-Type'='application/json',Accept='application/json'))

xmlmcURL <-  content(responseFromFiles, as = "parsed", type = "application/json", encoding="UTF-8")$zoneinfo$apiEndpoint

# invokeXmlmc - take params, fire off XMLMC call
invokeXmlmc = function(service, xmethod, params) {
  paramsrequest = paste(params , collapse="")
  arrRequest = c(	"<methodCall service=\"",
                  service, 
                  "\" method=\"",
                  xmethod,
                  "\">",
                  "<params>",
                  paramsrequest,
                  "</params>",
                  "</methodCall>")
  responseFromURL <- POST(paste(xmlmcURL, service, "?method=", xmethod, sep=""),
                          add_headers('Content-Type'='text/xmlmc',Accept='application/json', Authorization=paste('ESP-APIKEY', apiKey, sep=" ")),
                          use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                          body=paste(arrRequest, collapse=""))
  return(responseFromURL)
}

# Fire off API call to get measure data
measureResponse = invokeXmlmc("reporting", "measureGetInfo", paste("<measureId>", measureID, "</measureId>",
                                                                   "<returnMeasureValue>true</returnMeasureValue>",
                                                                   "<returnMeasureTrendData>true</returnMeasureTrendData>"))
runOutput <- content(measureResponse, encoding="UTF-8")

runStatus <- runOutput$"@status"

if (runStatus == FALSE || runStatus == "fail") {
  stop(runOutput$state$error)
} else {
  # Return trendValue table object, with flattened values within, to ensure dateRange.from and dateRange.to are properly represented in the table output
  dataframe <- as.data.frame(do.call(rbind, lapply(runOutput$params$trendValue, as.data.frame)))
}


PowerBIDataSource_HistoricReport.R
#Define Instance Details
instanceName = "yourinstanceid"
apiKey = "yourapikey"

# Define Report details
reportID = "1"
reportRunID = "21"
useXLSX <- FALSE # FALSE = the script will use the CSV output from your report; TRUE = the script will use the XLSX output from your report

# Settings for using XLSX report output
# You must have XLSX output enabled against the target report in your Hornbill instance to use these settings
deleteLocalXLSX <- FALSE  # FALSE = the downloaded XLSX file will remain on disk once the extract is complete; TRUE = the local XLSX file is deleted upon completion
xlsxLocalFolder <- ""     # Can be left blank, or specify a local folder to store the downloaded XLSX file into. Requires the postfixed / or \ depending on your OS

# Settings for using CSV report output, or character encoding conversion for XLSX repost output
csvEncoding <- "UTF-8"    # For Unicode byte translation issues in Power BI, try using ISO-8859-1 as the value for csvEncoding

# Define Proxy Details
proxyAddress <- NULL  # "127.0.0.1" - location of proxy
proxyPort <- NULL     # 8080 - proxy port
proxyUsername <- NULL # login details for proxy, if needed
proxyPassword <- NULL # login details for proxy, if needed
proxyAuth <- NULL     # "any" - type of HTTP authentication to use. Should be one of the following: basic, digest, digest_ie, gssnegotiate, ntlm, any.

# Import dependencies
library('httr')

# Set httr default timeout, defaults to 10 seconds
set_config( config( connecttimeout = 60 ) )

# Get Endpoint
responseFromFiles <- GET(paste("https://files.hornbill.com/instances", instanceName, "zoneinfo",sep="/"), 
                         use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                         add_headers('Content-Type'='application/json',Accept='application/json'))

xmlmcURL <-  content(responseFromFiles, as = "parsed", type = "application/json", encoding="UTF-8")$zoneinfo$apiEndpoint
davURL <-  content(responseFromFiles, as = "parsed", type = "application/json", encoding="UTF-8")$zoneinfo$davEndpoint
# invokeXmlmc - take params, fire off XMLMC call
invokeXmlmc = function(service, xmethod, params) {
  paramsrequest = paste(params , collapse="")
  arrRequest = c(	"<methodCall service=\"",
                  service, 
                  "\" method=\"",
                  xmethod,
                  "\">",
                  "<params>",
                  paramsrequest,
                  "</params>",
                  "</methodCall>")
  apiUrl = paste(xmlmcURL, service, "?method=", xmethod, sep="")
  responseFromURL <- POST(apiUrl,
                          add_headers('Content-Type'='text/xmlmc',Accept='application/json', Authorization=paste('ESP-APIKEY', apiKey, sep=" ")),
                          use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                          body=paste(arrRequest, collapse=""))
  return(responseFromURL)
}

# Get filenames from report ID and run ID 
reportRunResponse = invokeXmlmc("reporting", "reportRunGetStatus", paste("<runId>", reportRunID, "</runId>"))
runOutput <- content(reportRunResponse, encoding="UTF-8")
runStatus <- runOutput$"@status"

if (runStatus == FALSE || runStatus == "fail") {
  stop(runOutput$state$error)
} else {
  for (file in runOutput$params$files) {
    if (file$type == "xlsx" && useXLSX == TRUE) {
      reportLink = file$name
    } else if (file$type == "csv" && useXLSX == FALSE) {
      reportLink = file$name
    }
  }
  reportLinkLocal <- paste(xlsxLocalFolder, reportLink, sep="")
  reportContent <- GET(paste(davURL, "reports", reportID, URLencode(reportLink), sep="/"),
                       write_disk(reportLinkLocal, overwrite=TRUE),
                       use_proxy(proxyAddress, proxyPort, auth = proxyAuth, username = proxyUsername, password = proxyPassword),
                       add_headers('Content-Type'='text/dav', Authorization=paste('ESP-APIKEY ', apiKey, sep="")))
  if (useXLSX == TRUE) {
    library(readxl)
    write.csv(read_excel(reportLinkLocal),paste(reportLinkLocal, "csv", sep="."), row.names = FALSE)
    output <-read.csv(paste(reportLinkLocal, "csv", sep="."), encoding = csvEncoding, stringsAsFactors=FALSE)
  } else {
    output <- content(reportContent, as = "parsed", type = "text/csv", encoding = csvEncoding)
  }
  if (deleteLocalXLSX == TRUE && file.exists(reportLinkLocal)) {
    file.remove(reportLinkLocal)
  }
  if (file.exists(paste(reportLinkLocal, "csv", sep="."))) {
    file.remove(paste(reportLinkLocal, "csv", sep="."))
  }
}